File: src/Cxx/StructuredGrid/BlankPoint.cxx
#include <vtkActor.h>
#include <vtkDataSetMapper.h>
#include <vtkIdList.h>
#include <vtkMath.h>
#include <vtkNamedColors.h>
#include <vtkNew.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>
#include <vtkStructuredGrid.h>
#include <vtkStructuredGridGeometryFilter.h>
#include <vtkXMLPolyDataWriter.h>
#include <vtkXMLStructuredGridWriter.h>

int main(int, char*[])
{
  vtkNew<vtkNamedColors> colors;

  // Create a grid
  vtkNew<vtkStructuredGrid> structuredGrid;

  vtkNew<vtkPoints> points;

  unsigned int gridSize = 8;
  unsigned int counter = 0;
  // Create a 5x5 grid of points
  for (unsigned int j = 0; j < gridSize; j++)
  {
    for (unsigned int i = 0; i < gridSize; i++)
    {
      if (i == 3 && j == 3) // Make one point higher than the rest
      {
        points->InsertNextPoint(i, j, 2);
        std::cout << "The different point is number " << counter << std::endl;
      }
      else
      {
        // Make most of the points the same height
        points->InsertNextPoint(i, j, 0);
      }
      counter++;
    }
  }

  // Specify the dimensions of the grid
  structuredGrid->SetDimensions(gridSize, gridSize, 1);

  structuredGrid->SetPoints(points);

  structuredGrid->BlankPoint(27);
  structuredGrid->Modified();

  // Create a mapper and actor
  vtkNew<vtkDataSetMapper> gridMapper;
  gridMapper->SetInputData(structuredGrid);

  vtkNew<vtkActor> gridActor;
  gridActor->SetMapper(gridMapper);
  gridActor->GetProperty()->EdgeVisibilityOn();
  gridActor->GetProperty()->SetEdgeColor(colors->GetColor3d("Blue").GetData());

  // Create a renderer, render window, and interactor
  vtkNew<vtkRenderer> renderer;
  vtkNew<vtkRenderWindow> renderWindow;
  renderWindow->AddRenderer(renderer);
  renderWindow->SetWindowName("BlankPoint");

  vtkNew<vtkRenderWindowInteractor> renderWindowInteractor;
  renderWindowInteractor->SetRenderWindow(renderWindow);

  // Add the actor to the scene
  renderer->AddActor(gridActor);
  renderer->SetBackground(colors->GetColor3d("ForestGreen").GetData());

  // Render and interact
  renderWindow->Render();

  renderWindowInteractor->Start();

  return EXIT_SUCCESS;
}

File: src/Python/StructuredGrid/BlankPoint.py
#!/usr/bin/env python3

# noinspection PyUnresolvedReferences
import vtkmodules.vtkInteractionStyle
# noinspection PyUnresolvedReferences
import vtkmodules.vtkRenderingOpenGL2
from vtkmodules.vtkCommonColor import vtkNamedColors
from vtkmodules.vtkCommonCore import vtkPoints
from vtkmodules.vtkCommonDataModel import vtkStructuredGrid
from vtkmodules.vtkRenderingCore import (
    vtkActor,
    vtkDataSetMapper,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkRenderer
)


def main():
    colors = vtkNamedColors()

    structured_grid = vtkStructuredGrid()

    points = vtkPoints()

    grid_size = 8
    counter = 0
    # Create a 5x5 grid of points
    for j in range(0, grid_size):
        for i in range(0, grid_size):
            if i == 3 and j == 3:  # Make one point higher than the rest.
                points.InsertNextPoint(i, j, 2)
                print('The different point is number ', counter)
            else:
                # Make most of the points the same height.
                points.InsertNextPoint(i, j, 0)
            counter += 1

    # Specify the dimensions of the grid.
    structured_grid.SetDimensions(grid_size, grid_size, 1)

    structured_grid.SetPoints(points)

    structured_grid.BlankPoint(27)
    structured_grid.Modified()

    # Check.
    def is_visible(pt_num):
        if structured_grid.IsPointVisible(pt_num):
            return f'Point {pt_num} is visible.'
        else:
            return f'Point {pt_num} is not visible.'

    # Should not be visible.
    print(is_visible(27))
    # Should be visible.
    print(is_visible(7))

    # Create a mapper and actor.
    grid_mapper = vtkDataSetMapper()
    grid_mapper.SetInputData(structured_grid)

    grid_actor = vtkActor()
    grid_actor.SetMapper(grid_mapper)
    grid_actor.GetProperty().EdgeVisibilityOn()
    grid_actor.GetProperty().SetEdgeColor(colors.GetColor3d('Blue'))

    # Visualize
    renderer = vtkRenderer()
    ren_win = vtkRenderWindow()
    ren_win.AddRenderer(renderer)
    iren = vtkRenderWindowInteractor()
    iren.SetRenderWindow(ren_win)

    renderer.AddActor(grid_actor)
    renderer.SetBackground(colors.GetColor3d('ForestGreen'))

    # ren_win.SetSize(640, 480)
    ren_win.SetWindowName('BlankPoint')
    ren_win.Render()
    iren.Start()


if __name__ == '__main__':
    main()
